### PR TITLE
Disabling retry_persistent_connection in Tests

### PR DESCRIPTION
**Pull request by ShubhAgarwal-dev**
## What do these changes do?

It disables retry logic by default in tests.

## Are there changes in behavior for the user?

No

## Is it a substantial burden for the maintainers to support this?

No

## Related issue number

#9141 

## Checklist

- [X] I think the code is well written
- [ ] Unit tests for the changes exist
- [x] Add a new news fragment into the `CHANGES/` folder
- [X] Documentation reflects the changes
- [X] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`

### CODE REVIEW

File: aiohttp/client.py [hunk 1]
```diff
@@ -286,6 +287,7 @@ def __init__(
         max_line_size: int = 8190,
         max_field_size: int = 8190,
         fallback_charset_resolver: _CharsetResolver = lambda r, b: "utf-8",
+        test: bool = False,
```
Comment (Dreamsorcerer) [inline aiohttp/client.py]: When I said to use a private attribute only, I meant to not have a parameter here.

File: aiohttp/test_utils.py [hunk 2]
```diff
@@ -293,7 +293,7 @@ def __init__(  # type: ignore[misc]
         self._server = server
         if cookie_jar is None:
             cookie_jar = aiohttp.CookieJar(unsafe=True)
-        self._session = ClientSession(cookie_jar=cookie_jar, **kwargs)
+        self._session = ClientSession(cookie_jar=cookie_jar, test=True, **kwargs)
```
Comment (Dreamsorcerer) [inline aiohttp/test_utils.py]: And then just modify it here:
`self._session._retry_connection = False`

Then we'll avoid exposing it to users for now.

File: aiohttp/client.py [hunk 3]
```diff
@@ -367,6 +369,7 @@ def __init__(
 
         self._default_proxy = proxy
         self._default_proxy_auth = proxy_auth
+        self._test = test
```
Comment (Dreamsorcerer) [inline aiohttp/client.py]: I'd suggest `_retry_connection` as a more descriptive name.

File: aiohttp/client.py [hunk 4]
```diff
@@ -539,7 +542,9 @@ async def _request(
         try:
             with timer:
                 # https://www.rfc-editor.org/rfc/rfc9112.html#name-retrying-requests
-                retry_persistent_connection = method in IDEMPOTENT_METHODS
+                retry_persistent_connection = (
+                    False if self._test else method in IDEMPOTENT_METHODS
```
Comment (Dreamsorcerer) [inline aiohttp/client.py]: This can probably become:
```suggestion
                    self._retry_connection and method in IDEMPOTENT_METHODS
```

**Comment by ShubhAgarwal-dev** [comment]
Hi @Dreamsorcerer, I will make all the suggested changes. I saw many tests are using this ClientSession class, should I make these changes everywhere..?

Also, how to pass the Label/Backport Label Check?

Thanks for the inputs

**Comment by ShubhAgarwal-dev** [comment]
Hi @Dreamsorcerer, I have made all the suggested changes.

**Comment by Dreamsorcerer** [comment]
There are a fair number of tests which create a ClientSession directly, but this should cover most tests.

We just want a test now to validate the behaviour in tests.

**Comment by Dreamsorcerer** [comment]
> We just want a test now to validate the behaviour in tests.

I'll probably take a look at it over the weekend, but if you'd like to give it a try, then I think we want a test that uses the aiohttp_client fixture with an app that has an endpoint which closes the connection (probably `request.close()`) and then assert that the client only made one request.

**Comment by ShubhAgarwal-dev** [comment]
Ok, I will try to make a test like that. Can you point me to parts where I can take a look for inspiration while writing tests..?

**Comment by Dreamsorcerer** [comment]
Probably any test that uses aiohttp_client currently. The test should be added to test_test_utils.py

**Comment by ShubhAgarwal-dev** [comment]
Hi @Dreamsorcerer, I have added the required test, I think it should work

**Comment by Dreamsorcerer** [comment]
Thanks, I think that's close to what we want. I'll play with it tomorrow and make some tweaks.

**Comment by ShubhAgarwal-dev** [comment]
Cool

**Review by ShubhAgarwal-dev — COMMENTED** [review]
Please let me know if there is anything else which I can try for hacktoberfest.

**Comment by webknjaz** [comment]
@ShubhAgarwal-dev while you did successfully change the code, I have feedback regarding using Git, specifically commit message and PR titles and descriptions. This is a very important skill both for yourself and when collaborating on projects with others. Please, check out this collection of articles on the topic: https://gist.github.com/webknjaz/cb7d7bf62c3dda4b1342d639d0e78d79. Following those recommendations will improve the quality of your interactions with Git and git hostings by a lot! Hope this helps :)

Regarding more issues, check out https://github.com/aio-libs/aiohttp/issues?q=sort%3Aupdated-desc+is%3Aopen+label%3AHacktoberfest to see if there's anything you're able to understand the context of.
Additionally, we have a number of aiohttp dependencies in the same @aio-libs org — frozenlist, multidict, propcache, yarl. Those are smaller libraries where you may be able to find something to work on as well.

File: CHANGES/9141.misc.rst [hunk 5]
```diff
@@ -0,0 +1,3 @@
+Modified the `ClientSession` class by adding a private attribute `_retry_connection` of type `bool`, with a default value of `True` to the class.
+`retry_persistent_connection` in `ClientSession` is set to `False` if `_retry_connection` was `False`.
+In the session of `TestClient`, `ClientSession` attribute `_retry_connection` is changed to `False` -- by :user:`ShubhAgarwal-dev`.
```
Comment (webknjaz) [inline CHANGES/9141.misc.rst]: This must not use Markdown's single backticks. The markup of the document is RST and any non-role highlighting should use double backticks. For the linkable objects, respective roles must be added. Like
```rst
:data:`False`
```
or
```rst
:class:`bool`
```

Comment (ShubhAgarwal-dev) [reply CHANGES/9141.misc.rst]: Thanks. I will revise the RST file as per your suggestions.

**Comment by ShubhAgarwal-dev** [comment]
> @ShubhAgarwal-dev while you did successfully change the code, I have feedback regarding using Git, specifically commit message and PR titles and descriptions. This is a very important skill both for yourself and when collaborating on projects with others. Please, check out this collection of articles on the topic: https://gist.github.com/webknjaz/cb7d7bf62c3dda4b1342d639d0e78d79. Following those recommendations will improve the quality of your interactions with Git and git hostings by a lot! Hope this helps :)
> 
> Regarding more issues, check out https://github.com/aio-libs/aiohttp/issues?q=sort%3Aupdated-desc+is%3Aopen+label%3AHacktoberfest to see if there's anything you're able to understand the context of. Additionally, we have a number of aiohttp dependencies in the same @aio-libs org — frozenlist, multidict, propcache, yarl. Those are smaller libraries where you may be able to find something to work on as well.

Thanks for the comments. I will go through the articles and try to adopt the best practices. Once done, It will be my first-time contributing the changes to such a big open source project.

**Comment by Dreamsorcerer** [comment]
Alright, updated test should be good now. I also rewrote the change fragment to focus on what a user will want to know.

File: tests/test_test_utils.py [hunk 6]
```diff
@@ -10,7 +10,8 @@
 from yarl import URL
 
 import aiohttp
-from aiohttp import web
+from aiohttp import ServerDisconnectedError, web
```
Comment (Dreamsorcerer) [inline tests/test_test_utils.py]:
```suggestion
from aiohttp import web
```

**Comment by Dreamsorcerer** [comment]
@ShubhAgarwal-dev Would you mind following the above instructions to create a backport PR? Thanks.

**Comment by ShubhAgarwal-dev** [comment]
Sorry, I didn't know how to create backport PR. I would have searched how to create one If I knew this has to be any kind of special PR. I just followed the instructions on the CONTRIBUTIONS.rst file.

**Comment by ShubhAgarwal-dev** [comment]
Ok Cool I will do the same

**Comment by Dreamsorcerer** [comment]
I mean the instructions that were just posted. :P

**Comment by Dreamsorcerer** [comment]
https://github.com/aio-libs/aiohttp/pull/9396#issuecomment-2395571754

Depending on conflicts, sometimes the bot does it directly without human intervention.

**Pull request merged**